core/lib/server/pub.ts

```diff
@@ -601,6 +606,383 @@ export const deletePub = async (pubId: PubsId) =>
+		communityId
+	);
+	const fieldIds = fields.map(({ fieldId }) => fieldId);
+	if (!fieldIds.length) {
+		throw new Error(`No fields found for slugs: ${slugs.join(", ")}`);
+	}
+
+	const removed = await autoRevalidate(
+		trx
+			.deleteFrom("pub_values")
+			.where("pubId", "=", pubId)
+			.where("fieldId", "in", fieldIds)
+			.where("relatedPubId", "is not", null)
+			.returning("relatedPubId")
+	).execute();
+
+	return removed.map(({ relatedPubId }) => relatedPubId);
+};
+
+/**
+ * Replaces all relations for given field slugs with new relations.
+ * First removes all existing relations for the provided slugs, then adds the new relations.
+ *
+ * If the `relations` object is empty, this function does nothing.
+ */
+export const replacePubRelationsBySlug = async ({
+	pubId,
+	relations,
+	communityId,
+	trx = db,
+}: {
+	pubId: PubsId;
+	relations: Record<string, Omit<AddPubRelationsInput, "slug">[]>;
+	communityId: CommunitiesId;
+	trx?: typeof db;
+}) => {
+	if (!Object.keys(relations).length) {
+		return;
+	}
+
+	await maybeWithTrx(trx, async (trx) => {
+		const slugs = Object.keys(relations);
+
+		await removeAllPubRelationsBySlugs({ pubId, slugs, communityId, trx });
+
+		const relationsWithSlug = slugs.flatMap((slug) =>
+			relations[slug].map((relation) => ({ ...relation, slug }) as AddPubRelationsInput)
+		);
+
+		await upsertPubRelations({ pubId, relations: relationsWithSlug, communityId, trx });
+	});
+};
+
+export const updatePub = async ({
+	pubId,
+	pubValues,
+	communityId,
+	stageId,
+	continueOnValidationError,
+}: {
+	pubId: PubsId;
+	pubValues: CreatePubRequestBodyWithNullsNew["values"];
+	communityId: CommunitiesId;
+	stageId?: StagesId;
+	continueOnValidationError: boolean;
+}) => {
+	const result = await db.transaction().execute(async (trx) => {
+		// Update the stage if a target stage was provided.
+		if (stageId !== undefined) {
+			await autoRevalidate(
+				trx.deleteFrom("PubsInStages").where("PubsInStages.pubId", "=", pubId)
+			).execute();
+			await autoRevalidate(
+				trx.insertInto("PubsInStages").values({ pubId, stageId })
+			).execute();
+		}
+
+		const vals = Object.entries(pubValues).flatMap(([slug, value]) => ({
+			slug,
+			value,
+		}));
+
+		const pubValuesWithSchemaNameAndFieldId = await validatePubValues({
+			pubValues: vals,
+			communityId,
+			continueOnValidationError,
+		});
+
+		if (!pubValuesWithSchemaNameAndFieldId.length) {
+			return {
+				success: true,
+				report: "Pub not updated, no pub values to update",
+			};
+		}
+
+		try {
+			await autoRevalidate(
+				trx
+					.insertInto("pub_values")
+					.values(
+						pubValuesWithSchemaNameAndFieldId.map(({ value, fieldId }) => ({
+							pubId,
+							fieldId,
+							value: JSON.stringify(value),
+						}))
+					)
+					.onConflict((oc) =>
+						oc
+							// we have a unique index on pubId and fieldId where relatedPubId is null
+							.columns(["pubId", "fieldId"])
+							.where("relatedPubId", "is", null)
```

Comment:
very nice, i tried this before but couldn't get it to work since i was specifying the name of the index. i now know that you have to specify the columns and condition for postgres to find a partial index in an on conflict statement!

Reply:
lot of trial & error 😅